frame/uniques/src/types.rs

```diff
@@ -125,3 +131,40 @@ pub struct InstanceMetadata<DepositBalance, StringLimit: Get<u32>> {
 	/// Whether the asset metadata may be changed by a non Force origin.
 	pub(super) is_frozen: bool,
 }
+
+/// Represents either a System currency or a set of fungible assets.
+#[derive(Encode, Decode, Clone, PartialEq, MaxEncodedLen, RuntimeDebug, TypeInfo)]
+pub enum BalanceOrAsset<Balance, AssetId, AssetBalance> {
+	Balance { amount: Balance },
+	Asset { id: AssetId, amount: AssetBalance },
+}
+
+impl<B, A, AB> From<B> for BalanceOrAsset<B, A, AB> {
+	fn from(amount: B) -> Self {
+		Self::Balance { amount }
+	}
+}
+
+impl<B, A, AB> BalanceOrAsset<B, A, AB>
+where
+	A: PartialEq,
+	B: PartialOrd,
+	AB: PartialOrd,
+{
+	pub fn is_greater_or_equal(&self, other: &Self) -> bool {
+		use BalanceOrAsset::*;
+		match (self, other) {
+			(Balance { amount: a }, Balance { amount: b }) => a >= b,
+			(Asset { amount: a, .. }, Asset { amount: b, .. }) => a >= b,
+			_ => false,
```

Comment:
You could also return an `WrongCurrency` here and throw out the `is_same_currency`.

Reply:
Done